### PR TITLE
Add type hints to logging/context.py

### PR DESCRIPTION
**Pull request by neiljp**
This is a preliminary exploration of adding fuller type hints to a file that seems to be checked by mypy in CI - not all are? I can expand this PR to include more of `logging/` if this kind of approach is deemed reasonable.

I'm not running a full dev environment right now due to lack of space/time, but have a matching mypy installed. I have one outstanding type hint complaint by mypy from my own changes, and one that appears to be in upstream.

I've split the flake8 changes into a separate commit, for now, as I'm not sure if they're all necessary or desirable - partly based on my experience with type hint style in Zulip. See the comments in the commit - basically re spacing and quotes.

Most changes in the first commit should be straightforward, but I'm unsure regarding whether:
* extra assert statements are acceptable; in Zulip they're used to help with typing
* the removal of the `self.previous_context = None` line (currently commented); it being `None` complicates the typing and appears a bit strange in any case 

This does not currently fully add type hints to every function, but covers a good proportion.

### CODE REVIEW

**Comment by auscompgeek** [comment]
re: spacing around type hints for default args - this is a PEP 8 recommendation.

re: forward declaration within the same class - `from __future__ import annotations` isn't used (that's new in Python 3.7) so annotations need to be valid expressions at runtime.

**Comment by neiljp** [comment]
> re: spacing around type hints for default args - this is a PEP 8 recommendation.

Agreed; perhaps this was in flux when Zulip started using it or it was deemed too space-filling, though I know Zulip doesn't fully match PEP8 in other ways too.

> re: forward declaration within the same class - `from __future__ import annotations` isn't used (that's new in Python 3.7) so annotations need to be valid expressions at runtime.

Right, so to support Python 3.5 forward declarations are required.

File: changelog.d/6309.misc [hunk 1]
```diff
@@ -0,0 +1 @@
+Add type hints to logging/context.py.
```
Comment (anoadragon453) [inline changelog.d/6309.misc]:
```suggestion
Add type hints to `logging/context.py`.
```

**Comment by neiljp** [comment]
@hawkowl Any preliminary feedback on this would be useful :)

File: synapse/logging/context.py [hunk 2]
```diff
@@ -460,10 +478,10 @@ class LoggingContextFilter(logging.Filter):
             missing fields
     """
 
-    def __init__(self, **defaults):
+    def __init__(self, **defaults) -> None:
         self.defaults = defaults
 
-    def filter(self, record):
+    def filter(self, record) -> bool:  # FIXME: Use Literal[True] ?
```
Comment (anoadragon453) [inline synapse/logging/context.py]: `Literal[True]` works.

File: synapse/logging/context.py [hunk 3]
```diff
@@ -317,7 +333,7 @@ def __exit__(self, type, value, traceback):
                 logger.warning(
                     "Expected logging context %s but found %s", self, current
                 )
-        self.previous_context = None
+        # self.previous_context = None  # FIXME If this *can* be removed, this simplifies the typing...
```
Comment (anoadragon453) [inline synapse/logging/context.py]: As far as I can tell, this isn't used for anything, so feel free.

File: synapse/logging/context.py [hunk 4]
```diff
@@ -26,10 +26,13 @@
 import logging
 import threading
 import types
-from typing import Any, List
+from typing import Optional, Tuple, TypeVar, Union
 
 from twisted.internet import defer, threads
 
+if False:
+    from synapse.logging.scopecontextmanager import _LogContextScope
```
[the remark this answers is not in the document]
Comment (neiljp) [reply synapse/logging/context.py]: This avoids the import at runtime, but type checkers become aware of the name. In sufficiently recent versions this uses `typing.TYPE_CHECKING`, so I've changed the commit to use this.

File: synapse/logging/context.py (hunk 2 above)
Comment (neiljp) [reply synapse/logging/context.py]: I've not made this change yet, as only python 3.8 has `Literal` present, unless the project is willing to add a dependency upon `typing_extensions`.

Comment (richvdh) [reply synapse/logging/context.py]: fwiw, we already have a dependency on typing-extensions (currently on v3.7.4)

**Comment by neiljp** [comment]
@anoadragon453 Updated; should have covered all bases now.

**Comment by neiljp** [comment]
@anoadragon453 Just checking in, in case you're waiting on me here.

**Pull request merged**